compiler/ilgen/IlBuilder.cpp

```diff
@@ -1565,6 +1565,119 @@ IlBuilder::AtomicAdd(TR::IlValue * baseAddress, TR::IlValue * value)
+ *    +----------------------------------------+    +-------------------------------+    +-------------------------+
+ *                      |                                          |                                 |
+ *                      |                                          |                                 |
+ *                      |------------------------------------------+---------------------------------+
+ *                      |
+ *                      v             
+ *              +----------------+     
+ *              | block_$merge   |     
+ *              | ============== |   
+ *              |      ...       |
+ *              +----------------+   
+ *
+ * \endverbatim 
+ *
+ * \structure & \param value
+ *     tstart
+ *       |
+ *       ----persistentFailure // This is a permanent failure, try atomic way to do it instead
+ *       |
+ *       ----transientFailre // Temporary failure, user can retry 
+ *       |
+ *       ----fallthrough // Success, use general(non-atomic) way to do it
+ */
+void
+IlBuilder::TransactionBegin(TR::IlBuilder **persistentFailureBuilder, TR::IlBuilder **transientFailureBuilder, TR::IlBuilder **fallThroughBuilder)
+    {   
+    //ILB_REPLAY("%s->TransactionBegin(%s, %s, %s);", REPLAY_BUILDER(this), REPLAY_BUILDER(persistentFailureBuilder), REPLAY_BUILDER(transientFailureBuilder), REPLAY_BUILDER(fallThroughBuilder)); 
+    TraceIL("IlBuilder[ %p ]::transactionBegin %p, %p, %p, %p)\n", this, *persistentFailureBuilder, *transientFailureBuilder, *fallThroughBuilder);
+   
+    appendBlock();
+
+    TR::Block *mergeBlock = emptyBlock();
+    *persistentFailureBuilder = createBuilderIfNeeded(*persistentFailureBuilder);
+    *transientFailureBuilder = createBuilderIfNeeded(*transientFailureBuilder);
+    *fallThroughBuilder = createBuilderIfNeeded(*fallThroughBuilder);
+
+    TR::Node *persistentFailureNode = TR::Node::create(TR::branch, 0, (*persistentFailureBuilder)->getEntry()->getEntry());
+    TR::Node *transientFailureNode = TR::Node::create(TR::branch, 0, (*transientFailureBuilder)->getEntry()->getEntry());
+    TR::Node *fallThroughNode = TR::Node::create(TR::branch, 0, (*fallThroughBuilder)->getEntry()->getEntry());
+
+    TR::Node *tStartNode = TR::Node::create(TR::tstart, 3, persistentFailureNode, transientFailureNode, fallThroughNode);   
+    tStartNode->setSymbolReference(comp()->getSymRefTab()->findOrCreateTransactionEntrySymbolRef(comp()->getMethodSymbol()));
+    
+    genTreeTop(tStartNode);
+
+    //connecting the block having tstart with persistentFailure's and transientFailure's blocks 
+    cfg()->addEdge(_currentBlock, (*persistentFailureBuilder)->getEntry());
+    cfg()->addEdge(_currentBlock, (*transientFailureBuilder)->getEntry());
+    
+    AppendBuilder(*fallThroughBuilder);
+    gotoBlock(mergeBlock);
+
+    AppendBuilder(*persistentFailureBuilder);
+    gotoBlock(mergeBlock);
+    AppendBuilder(*transientFailureBuilder);
+    
```

Comment:
I think it would be better to lay out your "fallThroughBuilder" last so it can fall through to the mergeBlock. Following the tstart node, you should Goto the fallThroughBuilder (around the persistent and transient failure builders).

Why isn't the tend node automatically put at the end of the "fallThroughBuilder" ?